storage/remote/queue_manager.go

```diff
@@ -734,9 +707,10 @@ func (s *shards) sendSamples(ctx context.Context, samples []prompb.TimeSeries) {
 }
 
 // sendSamples to the remote storage with backoff for recoverable errors.
-func (s *shards) sendSamplesWithBackoff(ctx context.Context, samples []prompb.TimeSeries) error {
+func (s *shards) sendSamplesWithBackoff(ctx context.Context, samples []prompb.TimeSeries, buf *[]byte) error {
 	backoff := s.qm.cfg.MinBackoff
-	req, highest, err := buildWriteRequest(samples)
+	req, highest, err := buildWriteRequest(samples, *buf)
+	*buf = req
```

Comment:
Do you think this assignment to the pointer should be in `buildWriteRequest`?

Reply:
I put it here because it is the last place that deals with `*shards` and I wanted to keep the mutation isolated to those functions. I am happy to move it down to `buildWriteRequest` if you have a strong preference.